### PR TITLE
Meta: Always enable header check in serenity.sh

### PR DESCRIPTION
**Pull request by tomuta**
Took me way too long to figure out how `ENABLE_COMPILETIME_HEADER_CHECK` works. Turns out it works for everything except lagom. So let's just always enable it when using `Meta/serenity.sh`.

### CODE REVIEW

**Comment by linusg** [comment]
cc @BenWiederhake

**Comment by trflynn89** [comment]
Is there a significant build time impact to enabling this option?

**Comment by tomuta** [comment]
> Is there a significant build time impact to enabling this option?

About 20-25 seconds difference for me on a clean build (deleting `Build/` and running `ccache -C` prior). But since we fail CI if those additional checks fail, it feels like we should enable them. I just had this problem with #9280 where it took me a while to figure out that CI fails because it's doing additional stuff that my local build wasn't doing. And enabling those checks wasn't as straight forward as I had hoped.

**Comment by IdanHo** [comment]
Can we maybe build the binary, but not actually run it, and then run it in the pre-commit check? (which would result in extra 20 seconds per commit, instead of per build)

**Comment by tomuta** [comment]
> Can we maybe build the binary, but not actually run it, and then run it in the pre-commit check? (which would result in extra 20 seconds per commit, instead of per build)

This doesn't add 20 seconds to each build, just the first. It finishes extremely quickly once cached.

**Comment by awesomekling** [comment]
What *bugs* does this actually prevent? I'm concerned that we're slowly bloating the CI and pre-commit runtimes with low-value static checks.

**Comment by tomuta** [comment]
> What _bugs_ does this actually prevent? I'm concerned that we're slowly bloating the CI and pre-commit runtimes with low-value static checks.

Not sure. The only reason I created this PR is that I had my other PR and CI started failing at one point. Today I decided to finally get to the bottom of it, and it was much more painful than it should have. Neither running tests nor building failed on my machine, yet CI kept failing doing something I wasn't aware of. Turns out, it's because of `ENABLE_COMPILETIME_HEADER_CHECK`. Which I didn't realize wouldn't do anything for lagom builds. Once I figured out that that flag only works for regular builds I was finally able to do something about these CI failures. I don't know what these header checks do exactly, but it feels like if we generally want them to pass we maybe should enable those checks by default so people don't create PRs that will fail without a straight forward way to figure out why.

**Comment by BenWiederhake** [comment]
> What _bugs_ does this actually prevent? I'm concerned that we're slowly bloating the CI and pre-commit runtimes with low-value static checks.

I feel like this is the point in time where I should finally reply already :D

As explained in #10321:

> Imagine you're happily writing a .cpp file, and you start with something like this:
> 
> ```
> #include <LibFoo/Bar.h>
> // And nothing else
> ```
> 
> … and then you get a compilation error. That would be weird, wouldn't it?

So the bugs that are actually prevented by this are "weird compilation errors". Like those that were fixed by #10321, or those into which we are still running all the time, because auto-generated headers are missing some includes ([WindowServer](https://github.com/SerenityOS/serenity/issues/10907),  [WindowManagerServer](https://github.com/SerenityOS/serenity/issues/9584), and [something that depends on ShareableBitmap](https://discord.com/channels/830522505605283862/830522505605283866/909067304749383681)).

Also, to clear something up: "HeaderCheck" is not a binary that runs and checks the code. It's rather a piece of cmake/python that tells our build system to (basically) compile each header as a cpp file, and therefore this should be able to cache rather easily.

However, the added time, and because I thought that the typical PR doesn't introduce new `.h` files (and therefore is not typically likely to introduce such bugs), I'd turn the option off by default. Turns out, that was a bad call.

In summary, I approve the intent of this PR. I don't know whether it's more appropriate to do it in `serenity.sh` or in `common_options.cmake`:

```diff
-serenity_option(ENABLE_COMPILETIME_HEADER_CHECK OFF CACHE BOOL "Enable compiletime check that each library header compiles stand-alone")
+serenity_option(ENABLE_COMPILETIME_HEADER_CHECK ON CACHE BOOL "Enable compiletime check that each library header compiles stand-alone")
```

**Comment by tomuta** [comment]
Yeah I don't think we should enable it wholesale, which is why I just enabled it in the `Meta/serenity.sh` convenience script rather than changing the default.

**Review by BenWiederhake — APPROVED** [review]

**Comment by creator1creeper1** [comment]
`Meta/serenity.sh` is basically the default way to build the system. To quote the discord: "If the underlying issue is trouble figuring out how to turn it on / test it, then it seems a documentation PR should suffice". I agree. I also think that a mismatch between the local tests and the CI is certainly very confusing. As Andreas mentioned: This probably won't find any critical bugs, just minor inconveniences with regards to header inclusion. Maybe we could run this daily instead of in each PR so that PRs will not fail because of it, but if such an issue comes up, people will still be notified?
